File: 3789_minimum_cost_to_acquire_required_items/Solution.cs
// LeetCode 3789 - Minimum Cost To Acquire Required Items
// https://leetcode.com/problems/minimum-cost-to-acquire-required-items/

using System;

public class Solution {
    public long MinimumCost(int cost1, int cost2, int costBoth, int need1, int need2) {
        long a = (long)need1 * cost1 + (long)need2 * cost2;
        long b = (long)costBoth * Math.Max(need1, need2);
        int mn = Math.Min(need1, need2);
        long c = (long)costBoth * mn + (long)(need1 - mn) * cost1 + (long)(need2 - mn) * cost2;
        return Math.Min(a, Math.Min(b, c));
    }
}

File: 3789_minimum_cost_to_acquire_required_items/Solution.scala
// LeetCode 3789 - Minimum Cost To Acquire Required Items
// https://leetcode.com/problems/minimum-cost-to-acquire-required-items/

object Solution {
  def minimumCost(cost1: Int, cost2: Int, costBoth: Int, need1: Int, need2: Int): Long = {
    val a = need1.toLong * cost1 + need2.toLong * cost2
    val b = costBoth.toLong * math.max(need1, need2)
    val mn = math.min(need1, need2)
    val c = costBoth.toLong * mn + (need1 - mn).toLong * cost1 + (need2 - mn).toLong * cost2
    math.min(a, math.min(b, c))
  }
}
